firewood/src/merkle/node.rs

```diff
@@ -402,6 +407,99 @@ impl<T> PartialEq for EncodedNode<T> {
     }
 }
 
+#[derive(Debug)]
+struct PathWithBitsPrefix(Path);
+
+impl<'de> Deserialize<'de> for PathWithBitsPrefix {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        // TODO danlaine: This is slightly abusing the serde API.
+        // We don't know the number of bytes in the tuple until we've read the first u64.
+        // We should implement a custom serializer/deserializer for EncodedNode.
+        deserializer.deserialize_tuple(usize::MAX, TupleVisitor)
+    }
+}
+
+struct TupleVisitor;
+
+impl<'de> Visitor<'de> for TupleVisitor {
+    type Value = PathWithBitsPrefix;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("tuple")
+    }
+
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::SeqAccess<'de>,
+    {
+        let path_bits_len = seq
+            .next_element::<u64>()?
+            .ok_or(serde::de::Error::custom("missing length"))?;
+
+        if path_bits_len % BITS_PER_NIBBLE != 0 {
+            return Err(serde::de::Error::custom(
+                "path length is not a multiple of 4",
+            ));
+        }
+
+        // Add BITS_PER_NIBBLE so that if there are an odd number of nibbles,
+        // we need to read the final byte.
+        // e.g. if `path_bits_len` is 12 we need to read 2 bytes
+        // If we didn't add BITS_PER_NIBBLE `path_bytes_len` would be 1.
+        let path_bytes_len = (path_bits_len + BITS_PER_NIBBLE) / BITS_PER_BYTE;
```

Comment:
Is this read from an untrusted source? seems like this could overflow.

Reply:
Added a max path length check to prevent malicious serialized paths from causing huge mallocs. Not the most elegant fix but we're planning on rewriting the node serialization/deserialization eventually so I'm OK with this as a stopgap.